Migrate to custom containers and better elements scoping

timur27 opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/7674

#### Changes proposed in this Pull Request
This PR creates a new div wrapper with a unique class name for WooPayments payment elements and uses it to search for elements within them. In addition, the direct targeting of WooCommerce elements has been replaced with targeting our element first and then traversing by looking at an element from that point.

This PR also fixes compatibility issue Vendrify theme had with WooPayments.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

_Shortcode checkout_
* smoke tests
* paying with new & saved cards
* displaying multiple payment methods correctly
* toggling payment method depending on billing country (e.g. Bancontact available for Belgium falls back to card once we change the country)
* Stripe Link works as expected 
* No payment method creation request when billing data is missing
* No errors in the console when enabling other gateway (e.g. Stripe) and pay with its payment method while still having WooPayments gateways enabled and presented on the checkout page


_Add payment method & Pay for order_
* smoke tests


_Themes_
* confirm checkout works with Vendrify theme
* check with a couple of other themes

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
scope payment method selectors within the payment container

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -35,8 +35,9 @@ export const getSelectedUPEGatewayPaymentMethod = () => {
-		'li.wc_payment_method input.input-radio:checked, li.woocommerce-PaymentMethod input.input-radio:checked'
+	const paymentContainer = document.querySelector( '#payment' );
+	const radio = paymentContainer?.querySelector(
+		'input[name="payment_method"]:checked'
```
I considered switching from `input[name="payment_method"]` to just search for input of type `radio`, but after some thinking it seems that this type of implementation, while not relying on any IDs/classnames, might unexpectedly catch some unrelated radio inputs from within the payments container. Happy to hear your opinion on this in particular.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9782` or branch name `scope/payment-elements-selectors` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=scope/payment-elements-selectors) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 8a67751b026cb0fbbab6e8d31e2f50dd6070de98
- Build time: 2024-12-05 09:08:19 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
scope save payment method checkbox to payment elements and use name not ID for querying

Commit:
scope save payment method checkbox retrieval

Commit:
scope saved payment method selector and refactor

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -183,17 +185,16 @@ export const appendFraudPreventionTokenInputToForm = ( $form ) => {
 		paymentMethodType === 'card' || paymentMethodType === 'link'
-			? prefix + suffix
-			: prefix + '_' + paymentMethodType + suffix;
+			? 'wc-woocommerce_payments-payment-token-new'
+			: `wc-woocommerce_payments_${ paymentMethodType }-payment-token-new`;
```
This is a small refactor that I believe makes the code more readable by removing the prefix and suffix concepts

Commit:
create a new wrapper to wrap everything within WooPayments

Commit:
use wcpay-upe-form for class of the contaner to align with wcpay-upe-element

Commit:
tmp

Commit:
simplify getting selected pm and decouple from selectors

Commit:
add changelog entry

Commit:
improve shouldIncludeTerms function

Commit:
refactor the function responsible for checking if using saved payment token

Commit:
improve pm toggling for country

timur27 commented on client/checkout/classic/event-handlers.js:
```diff
@@ -92,6 +92,22 @@ jQuery( function ( $ ) {
+	$( document.body ).on( 'payment_method_selected', function () {
+		// Get selected payment method using WooCommerce's approach, see /legacy/js/frontend/checkout.js
+		const selectedPaymentMethod = $(
+			'.woocommerce-checkout input[name="payment_method"]:checked'
```
We're approaching things the same way in WooCommerce [`checkout.js`](https://github.com/Automattic/woocommerce/blob/security/plugins/woocommerce/client/legacy/js/frontend/checkout.js#L120)

frosso replied:
Since we're still leveraging WooCommerce Core selectors here, do we _need_ to listen to the change event to duplicate which payment method is selected?
I.e.: if we're relying on `.woocommerce-checkout input[name="payment_method"]:checked` here, could we just query for the same selector when we need to determine if a specific payment method is selected, instead of duplicating the information via the `checked` attribute on each `div.wcpay-upe-form`?

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -83,9 +58,11 @@ export const getHiddenBillingFields = ( enabledBillingFields ) => {
 };
 
-export const getUpeSettings = () => {
+export const getUpeSettings = ( paymentMethodType ) => {
```
I decided to add `paymentMethodType` here because the current idea behind `getUpeSettings` is that we run it for every gateway/payment method and it should theoretically check for every payment method type. So we let `shouldIncludeTerms` to scope by the `wcpay-upe-form` which has the payment method type equal to the one coming as a parameter here, and check from there

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -339,18 +326,17 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
-	const upeContainer = document.querySelector(
-		'.payment_method_woocommerce_payments_' + paymentMethodType
-	);
+	const upeContainer = upeElement.closest( 'li.wc_payment_method' );
```
Traversing up trying to find the first parent which will be ready to be used so that we can properly hide/show the gateway

timur27 commented on includes/class-wc-payments-checkout.php:
```diff
@@ -415,16 +415,25 @@ function () use ( $prepared_customer_data ) {
-			if ( ! empty( $this->gateway->get_description() ) ) : ?>
-				<p><?php echo wp_kses_post( $this->gateway->get_description() ); ?></p>
+			?>
+			<div class="wcpay-upe-form" 
```
This is the main wrapper, which should serve as the base for all DOM traversal moving forward. Ideally, we should aim to locate this container in the DOM and then traverse within it to find the relevant elements needed for our purposes

Commit:
search for save payment method checkbox dynamically

Commit:
avoid directly targeting Blocks element by acquiring the email from store selectors

Commit:
some cosmetics and nullchecks

Commit:
improve Stripe Link button handler DOM traversal

Commit:
improve toggling pm for country on shortcode checkout

Commit:
search for save payment method checkbox dynamically

Commit:
Revert "search for save payment method checkbox dynamically"

This reverts commit a77ae2e8f9fba66b7dcf6d50618fb43b639c86b6.

Commit:
search for save payment method checkbox

Commit:
create constant for parent payment form class on the FE

Commit:
some readability improvements and pm toggling fix

Commit:
move further to scoping by specific containers and refactor some parts

Commit:
fix tests

timur27 commented on client/checkout/classic/event-handlers.js:
```diff
@@ -239,71 +244,15 @@ jQuery( function ( $ ) {
-				] )
-		) {
-			return true;
+			const billingInput = upeElement
+				?.closest( 'form.checkout' )
+				?.querySelector( '#billing_country' );
```
This should be a bit faster and in line with how we discussed scoping by our elements instead of directly targeting non-WooPayments elements. I didn't include the down traversal, but I'm curious to hear your thoughts on this.

timur27 commented on client/checkout/classic/event-handlers.js:
```diff
@@ -239,71 +244,15 @@ jQuery( function ( $ ) {
-		}
-	}
-
-	function isBillingInformationMissing() {
```
moving to `upe.js` as it's clearly a helper function

frosso replied:
This is fine for now because it seems to be an existing problem - but the `client/checkout/utils/upe.js` seems to be used for both block-based and shortcode-based checkout.
You'll notice that your changes are affecting the bundle size of both shortcode and block-based checkout, even tho they're meant for the shortcode-based checkout only.

Eventually, it'd be good to separate utilities used only on shortcode vs utilities used only on blocks vs utilities used for both.

timur27 replied:
Ah yeah, I can remember this topic now that you were bringing it up some time ago. Thanks, that makes sense, I'll leave it here for now.

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -25,42 +30,23 @@ export const getTerms = ( paymentMethodsConfig, value = 'always' ) => {
-			selectedPaymentMethod = paymentMethodType;
-			break;
-		}
-	}
-
-	return selectedPaymentMethod;
+	return selectedGateway?.dataset.paymentMethodType || null;
```
`payment_method_selected` handler in `event-handlers.js` should do everything for us and we just need to return the gateway that is checked.

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -120,22 +115,23 @@ export const getUpeSettings = () => {
+		`#wc-${ paymentsForm.dataset.gatewayId }-new-payment-method`
+	);
+
+	return savePaymentMethodCheckbox?.checked || false;
```
selecting our own element, then looking for save payment method checkbox within that element, also taking the gateway id into account as in theory this function should support save payment method checkbox for e.g. sepa

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -183,17 +179,22 @@ export const appendFraudPreventionTokenInputToForm = ( $form ) => {
+		return false;
+	}
+
+	return newPaymentTokenInput !== null && ! newPaymentTokenInput.checked;
```
similar idea as in `shouldIncludeTerms`
additionally, avoid building the saved payment method selector by re-using the form's gateway id

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -279,12 +280,13 @@ export const getPaymentMethodTypes = ( paymentMethodType ) => {
  */
 export const getBlocksEmailValue = () => {
-	return document.getElementById( 'email' ).value;
+	const cartData = wp.data.select( 'wc/store/cart' ).getCustomerData();
+	return cartData?.billingAddress?.email || '';
```
I thought why not use store API instead of targeting the element?

frosso replied:
I think you might be getting possibly outdated data from the data store.
[The comment here states that it's coming from the DOM](https://github.com/Automattic/woocommerce-payments/blob/1af0ef92f475b1e401ba8c4b0475ddf44824ff7d/client/checkout/blocks/payment-processor.js#L117).

I'm unsure why it's set as a separate step - but it seems to be this way from the original implementation: https://github.com/Automattic/woocommerce-payments/blame/f63d75f2c2a83e8dce804fa0b248e86eefbeeeac/client/checkout/blocks/payment-processor.js#L110-L118

Looking at the call stack:
`getBlocksEmailValue()` is called from:
- [`setBillingAddress`](https://github.com/Automattic/woocommerce-payments/blob/1af0ef92f475b1e401ba8c4b0475ddf44824ff7d/client/checkout/blocks/payment-processor.js#L118-L120)
- [`setShippingAddress`](https://github.com/Automattic/woocommerce-payments/blob/1af0ef92f475b1e401ba8c4b0475ddf44824ff7d/client/checkout/blocks/payment-processor.js#L121C6-L123)

`setBillingAddress`/`setShippingAddress` are coming themselves from `useCustomerData()`.
And `useCustomerData()` is using those two methods from the `wc/store/cart` data store: https://github.com/Automattic/woocommerce-payments/blob/1af0ef92f475b1e401ba8c4b0475ddf44824ff7d/client/checkout/blocks/utils.js#L22-L26

So if you're now using the same `wc/store/cart` to set the values as the ones we just retrieved from the same data store - what are we doing, really? 🤔

timur27 replied:
Hah, this doesn't make sense indeed. I moved back to getting the value from the DOM in 8be97f00555d977c6bbad0c3a142682cc520772d

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -293,16 +295,20 @@ export const getBlocksEmailValue = () => {
  */
 export const blocksShowLinkButtonHandler = ( linkAutofill ) => {
-	const emailInput = document.getElementById( 'email' );
+	const upeContainer = document.querySelector( '.wcpay-payment-element' );
```
Scoping by blocks payment element we create

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -331,26 +337,106 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
+			);
+
+			const cardPaymentMethodInput = cardPaymentForm
+				?.closest( '.wc_payment_method' )
```
unfortunately, some mention of WC elements directly here, because we want to click on those elements

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -331,26 +337,106 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
+
+	// first name and last name are kinda special - we just need one of them to be at checkout
+	const name = `${
+		form.querySelector(
```
scoping by the form instead of looking through the whole document

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -25,42 +30,23 @@ export const getTerms = ( paymentMethodsConfig, value = 'always' ) => {
-	const radio = document.querySelector(
-		'li.wc_payment_method input.input-radio:checked, li.woocommerce-PaymentMethod input.input-radio:checked'
+	const selectedGateway = document.querySelector(
+		`.wcpay-upe-form[checked]`
```
The string template shouldn't be necessary here, since no string is being interpolated.
```suggestion
		'.wcpay-upe-form[checked]'
```

However, my concern is another - the `checked` attribute is reserved for `input` elements of type `radio` or `checkbox`.
In this scenario, `.wcpay-upe-form` is a `div`.

Since this is only used for the "classic"/"shortcode" checkout, could we instead just rely on the `input` with `name` of `payment_method`?

```js
document.querySelector('input[name="payment_method"][value*="woocommerce_payments"]:checked')
```

This should give us the selected payment method.

To get the name/type of the payment method, we can do:
```js
export const getSelectedUPEGatewayPaymentMethod = () => {
	const selectedGateway = document.querySelector(
		'input[name="payment_method"][value*="woocommerce_payments"]:checked'
	);
	if ( ! selectedGateway ) {
		return null;
	}

	// 'woocommerce_payments_affirm' => 'affirm'
	// 'woocommerce_payments_p24' -> 'p24'
	// 'woocommerce_payments' -> ''
	const paymentMethodType = selectedGateway.value
		// non-card elements are prefixed with `woocommerce_payments_*`
		.replace( 'woocommerce_payments_', '' )
		// the card element is just called `woocommerce_payments` - we need to account for variation in the name
		.replace( 'woocommerce_payments', '' );

	// if the string is empty, it's the card element
	return paymentMethodType || 'card';
};
```

timur27 replied:
Thanks for mentioning the `checked` attribute purpose for other types of elements 👍 
Regarding the usage of input element from WooCoommerce - I knew about this element but my assumption was we want to avoid targeting them as much as possible, that's why I went with our custom one which forced me to create some handling of updating selected payment method (relates to [another comment](https://github.com/Automattic/woocommerce-payments/pull/9782#discussion_r1862386154)). 

Another option would be to traverse from our div to the checked input.
I'd love to hear your thoughts about targeting Core html elements, thank you!

frosso replied:
I understand the worry about saving ourselves from potential changes from WooCommerce Core's elements or structure, which could make our implementation brittle.

However, the selector `.woocommerce-checkout input[name="payment_method"]:checked` (which targets WooCommerce Core elements) is already used in this implementation in the `payment_method_selected` listener.
I don't think that further wrapping this logic in custom handling necessarily isolates the code more effectively.

Traversing the `.wcpay-upe-form` to the `input` could be a middle-ground, but I don't think it makes it less brittle.

I was recommending using the `'input[name="payment_method"][value*="woocommerce_payments"]:checked'` selector because the `payment_method` inputs are indeed rendered by WC Core, but WC Core also relies on them to process the shortcode checkout.
It's unlikely for a theme to interfere with them.
Since there can only be one `input[name="payment_method"]` element with the `checked` attribute, I thought it was safe to use this selector.

Please let me know if I missed anything! :)

timur27 replied:
> However, the selector .woocommerce-checkout input[name="payment_method"]:checked (which targets WooCommerce Core elements) is already used in this implementation in the payment_method_selected listener

Well, that's actually a fair point. 

I removed the event handler and applied your suggestion in https://github.com/Automattic/woocommerce-payments/pull/9782/commits/e81e3a469071bcb109557de63e494e9529b1e8da.

I did that with all the comments, because they seem to be helpful. Also, I wanted to modify your suggestion by adding `card` instead of an empty string in the second replace, but upon analyzing it realized that it's actually less descriptive and in line with the first replace, so I kept things as they are while falling back to card in the return statement.

frosso commented on includes/class-wc-payments-checkout.php:
```diff
@@ -415,16 +415,25 @@ function () use ( $prepared_customer_data ) {
+			<div class="wcpay-upe-form" 
+				data-payment-method-type="<?php echo esc_attr( $this->gateway->get_stripe_id() ); ?>"
+				data-gateway-id="<?php echo esc_attr( $this->gateway->id ); ?>"
+				<?php checked( $this->gateway->chosen, true ); ?>
```
A `checked` attribute should not be a valid attribute on a `div` element. Noone is stopping you, but the [W3C validator](https://validator.w3.org/nu/#textarea) would flag it as invalid.

```html
<!DOCTYPE html>
<html lang="en-US">
<head>
    <meta charset="UTF-8">
    <title>Checkout – WooCommerce Payments Dev</title>
</head>

<body>
<div checked="true" class="some-div"><p>Some content</p></div>
</body>
</html>
```

<img width="615" alt="Screenshot 2024-11-28 at 4 55 04 PM" src="https://github.com/user-attachments/assets/b8eb608b-0b21-4527-9e05-4dfc55643ad9">

I think we should use valid markup on our merchant's sites.

timur27 replied:
Good point. I actually forgot to remove this redundant line because it was me testing how Core is setting the checked gateway on the page load. This is covered by WooCommerce and its input element (which we're currently actively discussing in another thread) so no need to include that here. Removed in fd1b17cf1f9a01627053e36942ddef6154e4892d

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -331,26 +337,106 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
+	// We need to just find one field with missing information. If even only one is missing, just return early.
+	return Boolean(
+		billingFieldsToValidate.find( ( fieldName ) => {
+			const $field = form.querySelector( `#${ fieldName }` );
```
I think this was pre-existing, but the `$` prefix is typically used for jQuery elements
```suggestion
			const field = form.querySelector( `#${ fieldName }` );
```

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -331,26 +337,106 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
+}
+
+export const isBillingInformationMissing = ( $form ) => {
+	const form = $form[ 0 ];
```
If we're needing just one `form` element, could we instead refactor `isBillingInformationMissing( form )` (and its only usage in `client/checkout/classic/event-handlers.js`) like this?
```js
// client/checkout/classic/event-handlers.js
	$checkoutForm.on( generateCheckoutEventNames(), function () {
		if ( isBillingInformationMissing( this ) ) {
			return;
		}
...
     } );

// client/checkout/utils/upe.js
export const isBillingInformationMissing = ( form ) => {
	const enabledBillingFields = getUPEConfig( 'enabledBillingFields' );
....
};
```

This way we don't even need to construct a jQuery element and then remove its jQuery wrapper around it?

timur27 replied:
Awesome, thanks so much for mentioning this <3 1c8549aa9dd11d6dcfde9b2ec968533e58836e82

frosso reviewed (commented):
You made some good improvements here 👍 

I have some questions about the validity of the markup and the selectors being used.

I also noticed this error at checkout with a brand new customer, with an empty form, changing from restricted to non-restricted countries (and vice-versa) for BNPL methos:
<img width="657" alt="Screenshot 2024-11-28 at 4 03 12 PM" src="https://github.com/user-attachments/assets/bb7f30fa-e5c7-415a-a774-e74c1ae6d920">

Commit:
get email value for Blocks from the DOM

Commit:
remove redundant checked attribute in theback-end

timur27 committed:
Update client/checkout/utils/upe.js

Co-authored-by: Francesco <frosso@users.noreply.github.com>

Commit:
follow up the $ prefix removal

Commit:
move away from jQuery

Commit:
remove pm change event handler and get the selected pm directly

Commit:
simplify pm toggling and fix tests

timur27 commented on client/checkout/utils/upe.js:
```diff
@@ -376,18 +376,11 @@ export const togglePaymentMethodForCountry = ( upeElement ) => {
-			const defaultPaymentMethod = 'card';
 			const cardPaymentForm = document.querySelector(
-				`.wcpay-upe-form[data-payment-method-type="${ defaultPaymentMethod }"]`
+				'input[name="payment_method"][value="woocommerce_payments"]'
```
cc @frosso I went back to targeting the Core input element in `togglePaymentMethodForCountry` as well in https://github.com/Automattic/woocommerce-payments/pull/9782/commits/05364e013c14434d5ead658d09e14024322447cc. It seems to simplify things

frosso commented:
> I also noticed this error at checkout with a brand new customer, with an empty form, changing from restricted to non-restricted countries (and vice-versa) for BNPL method

I can't reproduce this anymore 🤷 it's fine to dismiss it

Commit:
query billing country by name, not iD

timur27 commented:
@frosso thanks for your review so far. I've addressed all the opened threads and also did two additional things:

* Went back to [targeting the Core element while toggling payment method based on the country restrictions](https://github.com/Automattic/woocommerce-payments/pull/9782/files#r1863240174)
* Moved to [querying billing country element by name and not its ID](https://github.com/Automattic/woocommerce-payments/pull/9782/commits/618a90887d490b674f2edea72d0bc4411292c74a)


May I ask to have another look at the PR once your time allows? 👀

frosso commented on includes/class-wc-payments-checkout.php:
```diff
@@ -415,16 +415,24 @@ function () use ( $prepared_customer_data ) {
+			?>
+			<div class="wcpay-upe-form" 
+				data-payment-method-type="<?php echo esc_attr( $this->gateway->get_stripe_id() ); ?>"
+				data-gateway-id="<?php echo esc_attr( $this->gateway->id ); ?>"
```
Is it necessary to have both `data-payment-method-type` and `data-gateway-id`?

e.g.: when the payment method is `card`, we'll have `data-payment-method-type="card"` and `data-gateway-id="woocommerce_payments"`.
In case of `affirm`, we'll have `data-payment-method-type="affirm"` and `data-gateway-id="woocommerce_payments_affirm"`.

Could we just rely on one of those attributes?

timur27 replied:
We can do that I believe (7e7de87dd1766c2e5c7df586700402c11a117a44), although I was personally leaning toward passing it to avoid [string manipulations on the front-end](https://github.com/Automattic/woocommerce-payments/pull/9782/commits/7e7de87dd1766c2e5c7df586700402c11a117a44#diff-450d035a9e9ad47cb72bd13b3a7138e13cb0595c6a2a8c419b81524a77b10abdR134-R139). Applied the change, if you think this string manipulation is OK, let's keep this most recent change.

frosso replied:
Nice, thank you! There should similar references also in the tests you modified

frosso commented on includes/class-wc-payments-checkout.php:
```diff
@@ -435,19 +443,19 @@ function () use ( $prepared_customer_data ) {
-			?>
+				?>
 
 			<fieldset style="padding: 7px" id="wc-<?php echo esc_attr( $this->gateway->id ); ?>-upe-form" class="wc-upe-form wc-payment-form">
```
I find `wc-upe-form` and `wcpay-upe-form` named closely similar :D 

Since `wc-upe-form` is not used anywhere, could we remove it?

I wonder if we can just remove the `id` as well, at this point.

timur27 replied:
Great point, https://github.com/Automattic/woocommerce-payments/pull/9782/commits/7215501fb22ea3cb53685f3b2f0d4ed853fabb3b 👍

Commit:
do not pass gateway id to the FE

Commit:
cleanup the fieldset element

Commit:
fix tests

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -279,12 +305,16 @@ export const getPaymentMethodTypes = ( paymentMethodType ) => {
+	return document
+		.querySelector( '.wcpay-payment-element' )
+		?.closest( 'form' )
+		?.querySelector( 'input[type="email"]' )?.value;
```
Up to 3 months ago the `email` field didn't have `type="email"`: https://github.com/woocommerce/woocommerce/pull/48611/files

Should we use the previous fallback, where we were getting the email field by its `id`?

timur27 replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/9782/commits/4515b5156b7f4194a352ca6b1b6f0a5a6c0aa0b5

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -293,16 +323,20 @@ export const getBlocksEmailValue = () => {
+
+	const emailInput = upeContainer
+		.closest( 'form' )
+		?.querySelector( 'input[type="email"]' );
```
Same as above

Commit:
get billing email by ID in Blocks

Commit:
get billing email by ID in Blocks II

Commit:
fix tests

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -279,12 +305,16 @@ export const getPaymentMethodTypes = ( paymentMethodType ) => {
+	return document
+		.querySelector( '.wcpay-payment-element' )
+		?.closest( 'form' )
+		?.querySelector( '#billing_email' )?.value;
```
In the block-based checkout, I see that the email input has `id` attribute of `email`. I don't see a `billing_email` input.
<img width="1020" alt="Screenshot 2024-12-02 at 10 29 49 AM" src="https://github.com/user-attachments/assets/b8536b27-c039-4c66-b166-42749d2fafc4">

timur27 replied:
I confused this with the shortcode checkout elements, sorry about that. Moved to the correct ID in 70d2f84f735d7a9615a480081ea3e68378354a82

Commit:
fix the id of the billing email for blocks

frosso commented on client/checkout/classic/event-handlers.js:
```diff
@@ -251,71 +240,15 @@ jQuery( function ( $ ) {
-			return true;
+			const billingInput = upeElement
+				?.closest( 'form.checkout' )
+				?.querySelector( '[name="billing_country"]' );
```
```suggestion
				?.querySelector(
					`[name="${ SHORTCODE_BILLING_ADDRESS_FIELDS.country }"]`
				);
```

Very minor suggestion, since I noticed that the previous changes were making references to the `SHORTCODE_BILLING_ADDRESS_FIELDS` constant

frosso commented on client/checkout/utils/upe.js:
```diff
@@ -183,17 +202,24 @@ export const appendFraudPreventionTokenInputToForm = ( $form ) => {
+		return false;
+	}
+
+	return newPaymentTokenInput !== null && ! newPaymentTokenInput.checked;
```
```suggestion
	return ! newPaymentTokenInput.checked;
```

Since `! newPaymentTokenInput` is already being checked above, we shouldn't need to `!== null`

frosso reviewed (approved):
This looks good on my end, I don't see any blocking issues. Thank you for all the changes! :shipit:

Commit:
address the most recent comments

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

timur27 commented:
After the reviews so far, I thought I would attach the most recent testing round results I performed with the Vendrify theme. 



https://github.com/user-attachments/assets/baf21ab6-ff68-4b66-aa59-3ac2f2851400

mdmoore reviewed (approved):
Everything is testing well with the changes and checkouts are processing as expected. Checkout is working on Vendrify, but a couple things (which I'm not sure if are meant to be fixed in this PR) stand out. 

**Vendrify theme only**
1. Dynamic payment method filtering isn't working during checkout. Only the payment methods supported by the country set at page load are displayed and changing the billing country has no effect on the payment methods.
2. Pay for Order - I think is beyond our control but surfacing anyways. 
  a. If WooPayments is active, it's always displayed as the top payment method, regardless of payment method ordering at **WooCommerce > Payments**
  b. All other payment methods are hidden except for bullets. This includes WooPayments LPMs. 

![image](https://github.com/user-attachments/assets/e874b5ab-2199-4303-83bc-f965c66bb24f)

With the above in mind, since checkout is working with the new scoping and containers and checkout in Vendrify is now working, I think we can approve this and look at the above isssues separately if needed.

timur27 commented:
Thanks for the review and such a thorough testing, @mdmoore! 🚀  There are things with that theme that could work better, I agree. I think it would be OK to focus on fixing the critical flow, which is performing a purchase, and keep the rest part of the particular theme implementation (ideally) or as a separate issue (less preferred because I think there are limits to what we can do in terms of theme compatibility, as fixing one theme might break others). By the way, this theme is not being maintained anymore, from what I understand. 

The dynamic payment method filtering issue might be related to https://github.com/Automattic/woocommerce-payments/issues/9851, which I opened last week. It occurs even with some default testing themes. Maybe fixing this issue will fix the issue you encountered, but I'd focus first on the primary behavior and then on any potential user reports about problems with the theme.

timur27 committed:
Merge branch 'develop' into scope/payment-elements-selectors

timur27 commented:
Performed another small round of testing with fresh eyes today.

#### Shortcode
* check if billing information is missing upon checkout processing start ✅
* record user action for place order button click on the checkout page ✅
* payment elements mounting with Twenty Twenty-Four, Hello Elementor, Storefront, Electornics Retailer, Vendrify ✅ 
* payment method toggling based on location ✅
* payment form and save payment method checkbox retrieval for card, bancontact and p24 ✅
* `isUsingSavedPaymentMethod` function with card and p24 ✅
* toggling payment method for country ✅ 
   * falling back to CC if the currently selected payment method should be toggled off due to country restrictions ✅ 

#### Blocks 
* showing Link button handler within the email input element on Blocks ✅ 
* [getting email input value in Blocks](https://github.com/Automattic/woocommerce-payments/pull/9782/files#diff-450d035a9e9ad47cb72bd13b3a7138e13cb0595c6a2a8c419b81524a77b10abdR313-R317) (used for Stripe Link) ✅
